xcm/src/v1/multilocation.rs

```diff
@@ -36,83 +39,385 @@ use parity_scale_codec::{self, Decode, Encode};
+	}
+
+	/// Decrements the parent count by 1.
+	pub fn pop_parent(&mut self) {
+		self.parents = self.parents.saturating_sub(1);
+	}
+
+	/// Removes the first interior junction from `self`, returning it
+	/// (or `None` if it was empty or if `self` contains only parents).
+	pub fn take_first_interior(&mut self) -> Option<Junction> {
+		self.interior.take_first()
+	}
+
+	/// Removes the last element from `interior`, returning it (or `None` if it was empty or if
+	/// `self` only contains parents).
+	pub fn take_last(&mut self) -> Option<Junction> {
+		self.interior.take_last()
+	}
+
+	/// Ensures that `self` has the same number of parents as `prefix`, its junctions begins with
+	/// the junctions of `prefix` and that it has a single `Junction` item following.
+	/// If so, returns a reference to this `Junction` item.
+	///
+	/// # Example
+	/// ```rust
+	/// # use xcm::v1::{Junctions::*, Junction::*, MultiLocation};
+	/// # fn main() {
+	/// let mut m = MultiLocation::new(1, X2(PalletInstance(3), OnlyChild)).unwrap();
+	/// assert_eq!(
+	///     m.match_and_split(&MultiLocation::new(1, X1(PalletInstance(3))).unwrap()),
+	///     Some(&OnlyChild),
+	/// );
+	/// assert_eq!(m.match_and_split(&MultiLocation::new(1, Null).unwrap()), None);
+	/// # }
+	/// ```
+	pub fn match_and_split(&self, prefix: &MultiLocation) -> Option<&Junction> {
+		if self.parents != prefix.parents {
+			return None
+		}
+		self.interior.match_and_split(&prefix.interior)
+	}
+
+	/// Mutate `self` so that it is suffixed with `suffix`. The correct normalized form is returned,
+	/// removing any internal [Non-Parent, `Parent`]  combinations.
+	///
+	/// Does not modify `self` and returns `Err` with `suffix` in case of overflow.
+	///
+	/// # Example
+	/// ```rust
+	/// # use xcm::v1::{Junctions::*, Junction::*, MultiLocation};
+	/// # fn main() {
+	/// let mut m = MultiLocation::new(1, X2(Parachain(21), OnlyChild)).unwrap();
+	/// assert_eq!(m.append_with(MultiLocation::new(1, X1(PalletInstance(3))).unwrap()), Ok(()));
+	/// assert_eq!(m, MultiLocation::new(1, X2(Parachain(21), PalletInstance(3))).unwrap());
+	/// # }
+	/// ```
+	pub fn append_with(&mut self, suffix: MultiLocation) -> Result<(), MultiLocation> {
+		let mut prefix = suffix;
+		core::mem::swap(self, &mut prefix);
+		match self.prepend_with(prefix) {
+			Ok(()) => Ok(()),
+			Err(prefix) => {
+				let mut suffix = prefix;
+				core::mem::swap(self, &mut suffix);
+				Err(suffix)
+			},
+		}
+	}
+
+	/// Mutate `self` so that it is prefixed with `prefix`.
+	///
+	/// Does not modify `self` and returns `Err` with `prefix` in case of overflow.
+	///
+	/// # Example
+	/// ```rust
+	/// # use xcm::v1::{Junctions::*, Junction::*, MultiLocation};
+	/// # fn main() {
+	/// let mut m = MultiLocation::new(2, X1(PalletInstance(3))).unwrap();
+	/// assert_eq!(m.prepend_with(MultiLocation::new(1, X2(Parachain(21), OnlyChild)).unwrap()), Ok(()));
+	/// assert_eq!(m, MultiLocation::new(1, X1(PalletInstance(3))).unwrap());
```

Comment:
Can you elaborate this example? This is not what I intuitively would imagine *prefixing* looks like.

Reply:
Perhaps rewriting the MultiLocation format in a different way would make it more apparent:

`m` originally is in the form `../../PalletInstance(3)`, and now we want to prepend `../Parachain(21)/OnlyChild` to `m`. Or in other words, we start with `../Parachain(21)/OnlyChild` and we append `../../PalletInstance(3)` to it. We see that `m` contains 2 parent junctions, which will cause the `Parachain(21)/OnlyChild` part to be "eliminated" (since we are going back 2 "directories"/junctions up), and now we append `PalletInstance(3)` to the result, which would bring us to `../PalletInstance(3)`.